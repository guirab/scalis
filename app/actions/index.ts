"use server";
export async function create(formData: NewAccountType) {
  return (
    await fetch("http://localhost:3000/api/accounts", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(formData),
  }))?.json();
}

export async function getAll() {
  return (
    await fetch("http://localhost:3000/api/accounts")
  )?.json();
}

export async function login(formData: LoginType) {
  return await fetch("http://localhost:3000/api/login", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(formData),
  }).then((res:any) => {return res.json()});
}

export async function transferToSameAcc({action, type, amount, id, from}: UpdateType) {
  return await fetch(`http://localhost:3000/api/account/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({action, type, amount, from}),
  }).then((res:any) => {return res.json()});
}

export async function transferToOtherAcc({amount, to, password, id}: TransferOtherAccType) {
  return await fetch(`http://localhost:3000/api/account/transfer/${id}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ amount, to, password }),
  }).then((res:any) => {return res.json()});
}