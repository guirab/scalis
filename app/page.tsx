import Accounts from "./components/accounts";
import NewAccForm from "./components/newAccForm";

export default function Home() {
  return (
    <main className="flex min-h-screen flex-col items-center p-12 sm:p-24 justify-center sm:justify-start">
      <h1 className="text-5xl">Banking App</h1>
      <div className="w-full flex justify-between flex-col sm:flex-row gap-y-4">
        <NewAccForm />
        <Accounts />
      </div>
    </main>
  );
}
