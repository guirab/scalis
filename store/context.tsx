"use client";

import React, { ReactNode, createContext, useState } from "react";

export const AccountsContext: React.Context<ContextType> = createContext(
  {} as ContextType
);

export const AccountsProvider = ({ children }: { children: ReactNode }) => {
  const [account, setAccount] = useState<AccountType>({} as AccountType);
  const [accounts, setAccounts] = useState<AccountType[]>([] as AccountType[]);

  return (
    <AccountsContext.Provider
      value={{
        account,
        setAccount,
        accounts,
        setAccounts,
      }}
    >
      {children}
    </AccountsContext.Provider>
  );
};
