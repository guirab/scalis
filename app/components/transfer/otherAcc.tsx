"use client";
import { useContext, useEffect, useState } from "react";

import { InputCurrency } from "../inputCurrency";
import { AccountsContext } from "../../../store/context";
import { getAll, transferToOtherAcc } from "../../actions";

export const TransferOtherAcc = ({ setOpen }: ActionCardType) => {
  const [amount, setAmount] = useState("");
  const [to, setTo] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const [accountsOptions, setAccountsOptions] = useState<SelectType[]>([]);

  const { accounts, account, setAccount } = useContext(AccountsContext);

  async function loadAccounts() {
    let options: any = [];
    if (accounts && accounts.length !== 0) {
      options = accounts.map((account: AccountType) => ({
        value: account.username,
      }));
    } else {
      const data = await getAll();
      options = data?.map((account: AccountType) => ({
        value: account.username,
      }));
    }
    setAccountsOptions(options);
  }

  useEffect(() => {
    loadAccounts();
  }, [accounts]);

  async function onClick() {
    if (account.checking < parseFloat(amount.replace(/[^0-9.]/g, ""))) {
      setError("Insufficient funds");
      return;
    }
    await transferToOtherAcc({
      amount: parseFloat(amount.replace(/[^0-9.]/g, "")),
      to: to || accountsOptions[0]?.value,
      password,
      id: account.id,
    })
      .then((data) => {
        if (data.error) {
          setError(data.error);
        } else {
          setAccount(data as AccountType);
          setError("");
          setOpen(false);
        }
      })
      .catch((error) => {
        console.error("Error updating account:", error);
      });
  }

  return (
    <div data-testid="transfer-other-acc">
      <div>
        <label htmlFor="other">To: &nbsp;</label>
        <select
          name="other"
          id="other"
          value={to || accountsOptions[0]?.value}
          onChange={(e) => setTo(e.target.value)}
          className="text-black pl-2 outline-none rounded-md w-full"
        >
          {accountsOptions.map((option) => (
            <option key={option.value} value={option.value}>
              {option.value}
            </option>
          ))}
        </select>
      </div>

      <label htmlFor="other">Password: &nbsp;</label>
      <input
        autoComplete="off"
        type="password"
        name="password"
        id="password"
        value={password}
        onChange={(e) => setPassword(e.target.value)}
        required
        className="text-black pl-2 outline-none rounded-md w-full"
      />
      <div>
        <label htmlFor="currency-other">Amount: &nbsp;</label>
        <InputCurrency
          value={amount}
          setValue={setAmount}
          name="currency-other"
          id="currency-other"
          className="text-black pl-2 appearance-none outline-none rounded-md"
        />
      </div>
      {error && <span className="text-red-500">{error}</span>}
      <button
        className="bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
        onClick={onClick}
      >
        Transfer
      </button>
    </div>
  );
};
