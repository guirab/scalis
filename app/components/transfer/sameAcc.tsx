"use client";
import { useContext, useState } from "react";

import { InputCurrency } from "../inputCurrency";
import { AccountsContext } from "../../../store/context";
import { transferToSameAcc } from "../../actions";

export const TransferSameAcc = ({ setOpen }: ActionCardType) => {
  const [amount, setAmount] = useState("");
  const [type, setType] = useState("checking");
  const [error, setError] = useState("");

  const { account, setAccount } = useContext(AccountsContext);

  const optionsAccount = [
    {
      label: "Checking",
      value: "checking",
    },
    {
      label: "Savings",
      value: "savings",
    },
  ];

  async function onClick() {
    if (
      account[type as UpdateType["type"]] <
      parseFloat(amount.replace(/[^0-9.]/g, ""))
    ) {
      setError("Insufficient funds");
      return;
    }

    const payload: any = {
      action: "transfer",
      type: type as UpdateType["type"],
      amount: parseFloat(amount.replace(/[^0-9.]/g, "")),
      id: account.id,
    };

    await transferToSameAcc({
      action: "transfer",
      type: type as UpdateType["type"],
      amount: parseFloat(amount.replace(/[^0-9.]/g, "")),
      id: account.id,
      from: type as UpdateType["type"],
    })
      .then((data) => {
        if (data.error) {
          setError(data.error);
        } else {
          setAccount(data as AccountType);
          setError("");
          setOpen(false);
        }
      })
      .catch((error) => {
        console.error("Error updating account:", error);
      });
  }

  return (
    <div>
      <div>
        <label htmlFor="same">From: &nbsp;</label>
        <select
          name="same"
          id="same"
          value={type}
          onChange={(e) => setType(e.target.value)}
          className="text-black pl-2 outline-none rounded-md w-full"
        >
          {optionsAccount.map((option) => (
            <option key={option.value} value={option.value}>
              {option.label}
            </option>
          ))}
        </select>
      </div>
      <label htmlFor="same">Into: &nbsp;</label>
      <select
        name="same"
        id="same"
        value={
          optionsAccount.filter((option) => option.value !== type)[0].value
        }
        onChange={(e) => setType(e.target.value)}
        className="text-black pl-2 outline-none rounded-md w-full"
        disabled
      >
        {optionsAccount.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>

      <div>
        <label htmlFor="currency-same">Amount: &nbsp;</label>
        <InputCurrency
          value={amount}
          setValue={setAmount}
          name="currency-same"
          id="currency-same"
          className="text-black pl-2 appearance-none outline-none rounded-md"
        />
      </div>
      {error && <span className="text-red-500">{error}</span>}
      <button
        className="bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
        onClick={onClick}
      >
        Transfer
      </button>
    </div>
  );
};
