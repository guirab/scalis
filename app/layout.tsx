import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";

import { AccountsProvider } from "@/store/context";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Banking App",
  description: "Banking App for SCALIS",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <AccountsProvider>{children}</AccountsProvider>
      </body>
    </html>
  );
}
