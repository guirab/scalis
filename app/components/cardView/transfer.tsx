"use client";
import { useState } from "react";

import { TransferSameAcc } from "../transfer/sameAcc";
import { TransferOtherAcc } from "../transfer/otherAcc";

export const TransferView: React.FC<React.HTMLProps<any> & ActionCardType> = ({
  open,
  setOpen,
  ...props
}) => {
  const [isSameAccount, setIsSameAccount] = useState("true");

  const currentView = {
    true: <TransferSameAcc open={open} setOpen={setOpen} />,
    false: <TransferOtherAcc open={open} setOpen={setOpen} />,
  }[isSameAccount];

  return (
    <>
      {open && (
        <div className="block w-full mt-4 space-y-4" {...props}>
          <div className="gap-x-2 mb-4 flex text-sm justify-between">
            <div className="flex justify-start items-center">
              <label className="pr-px" htmlFor="same">
                This account
              </label>
              <input
                data-testid="same-account"
                type="radio"
                id="same"
                value="same"
                checked={isSameAccount === "true"}
                onChange={() => setIsSameAccount("true")}
              />
            </div>
            <div className="flex justify-start items-center">
              <label className="pr-px" htmlFor="other">
                Other account
              </label>
              <input
                data-testid="other-account"
                type="radio"
                id="other"
                value="other"
                checked={isSameAccount === "false"}
                onChange={() => setIsSameAccount("false")}
              />
            </div>
          </div>
          {currentView}
        </div>
      )}
    </>
  );
};
