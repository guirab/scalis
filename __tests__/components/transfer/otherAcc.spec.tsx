import {
  render,
  screen,
  fireEvent,
  waitFor,
  cleanup,
} from "@testing-library/react";
import { TransferOtherAcc } from "../../../app/components/transfer/otherAcc";
import { AccountsContext } from "../../../store/context";
import { getAll, transferToOtherAcc } from "../../../app/actions";
import { afterEach, beforeEach, describe, expect, test, vitest } from "vitest";

global.fetch = vitest.fn();

function createFetchResponse(data: any) {
  return { json: () => new Promise((resolve) => resolve(data)) };
}

describe("TransferOtherAcc", () => {
  afterEach(() => {
    cleanup();
  });

  test("snapshot", () => {
    const { container } = render(<TransferOtherAcc open setOpen={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  test("loads accounts on mount", async () => {
    render(<TransferOtherAcc open setOpen={() => {}} />);
    const accountsResponse = [{ id: 1, username: "test" }];

    (fetch as any).mockResolvedValue(createFetchResponse(accountsResponse));

    const fetchedAccounts = await getAll();

    expect(fetch).toHaveBeenCalledWith("http://localhost:3000/api/accounts");
    expect(fetchedAccounts).toStrictEqual(accountsResponse);
  });

  // test("transfers funds to another account", async () => {
  //   const setOpen = vitest.fn();
  //   const setAccount = vitest.fn();
  //   const setAccounts = vitest.fn();
  //   const accounts: any = [
  //     {
  //       id: 1,
  //       checking: 200,
  //       savings: 300,
  //     },
  //   ];
  //   const account: any = {
  //     id: 1,
  //     checking: 200,
  //     savings: 300,
  //   };
  //   render(
  //     <AccountsContext.Provider
  //       value={{ account, setAccount, accounts, setAccounts }}
  //     >
  //       <TransferOtherAcc open setOpen={setOpen} />
  //     </AccountsContext.Provider>
  //   );

  //   fireEvent.change(screen.getByTestId("to-select"), {
  //     target: { value: "account2" },
  //   });
  //   fireEvent.change(screen.getByTestId("password-input"), {
  //     target: { value: "password" },
  //   });
  //   fireEvent.change(screen.getByTestId("currency-input"), {
  //     target: { value: "50" },
  //   });

  //   fireEvent.click(screen.getByText("Transfer"));

  //   (fetch as any).mockResolvedValue(
  //     createFetchResponse({
  //       id: 1,
  //       username: "test",
  //       checking: 50,
  //     })
  //   );

  //   const createData = await transferToOtherAcc({
  //     amount: 50,
  //     to: "account2",
  //     password: "password",
  //     id: 1,
  //   });

  //   expect(fetch).toHaveBeenCalledWith(
  //     `http://localhost:3000/api/account/transfer/1`,
  //     {
  //       method: "POST",
  //       headers: {
  //         "Content-Type": "application/json",
  //       },
  //       body: JSON.stringify({
  //         amount: 50,
  //         to: "account2",
  //         password: "password",
  //         id: 1,
  //       }),
  //     }
  //   );
  //   // expect(createData).toStrictEqual(transferData);
  //   // });
  // });

  test("displays error message when there are insufficient funds", () => {
    const setOpen = vitest.fn();
    render(<TransferOtherAcc open setOpen={setOpen} />);

    vitest.mock("react", async () => {
      return {
        ...(await vitest.importActual("react")),
        useContext: () => ({
          account: { id: 1, username: "test", checking: 10 },
        }),
      };
    });

    const fromSelect = screen.getByTestId("to-select");
    const amountInput = screen.getByTestId("currency-input");
    const transferButton = screen.getByRole("button");

    fireEvent.change(fromSelect, { target: { value: "checking" } });
    fireEvent.change(amountInput, { target: { value: "10000" } });
    fireEvent.click(transferButton);

    const errorMessage = screen.getByText("Insufficient funds");
    expect(errorMessage).not.toBeNull();
  });
});
