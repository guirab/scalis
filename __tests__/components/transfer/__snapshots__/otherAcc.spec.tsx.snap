// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransferOtherAcc > snapshot 1`] = `
<div>
  <div
    data-testid="transfer-other-acc"
  >
    <div>
      <label
        for="other"
      >
        To:  
      </label>
      <select
        class="text-black pl-2 outline-none rounded-md w-full"
        data-testid="to-select"
        id="other"
        name="other"
      />
    </div>
    <label
      for="other"
    >
      Password:  
    </label>
    <input
      autocomplete="off"
      class="text-black pl-2 outline-none rounded-md w-full"
      data-testid="password-input"
      id="password"
      name="password"
      required=""
      type="password"
      value=""
    />
    <div>
      <label
        for="currency-other"
      >
        Amount:  
      </label>
      <div
        class="relative w-full"
      >
        <span
          class="absolute text-black left-px"
        >
          $ 
        </span>
        <input
          autocomplete="off"
          class="text-black pl-4 text-black pl-2 appearance-none outline-none rounded-md w-full"
          data-testid="currency-input"
          id="currency-other"
          name="currency-other"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
    >
      Transfer
    </button>
  </div>
</div>
`;
